File: backend/app/schemas/receipt.py
"""Receipt DTO (M11) — a print/display-friendly projection of a bill.

Store header + footer come from the M10 settings KV; the money breakdown is
the snapshot already stored on the Order. Voided lines are omitted (they
aren't on the customer's bill).
"""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel

from app.models.order import OrderChannel, OrderStatus
from app.models.payment import PaymentMethod


class ReceiptStore(BaseModel):
    name: str
    address: str | None
    tax_id: str | None


class ReceiptModifier(BaseModel):
    name: str
    price_delta: Decimal


class ReceiptLine(BaseModel):
    product_name: str
    qty: int
    unit_price: Decimal
    modifiers: list[ReceiptModifier]
    line_total: Decimal  # gross: qty * (unit_price + Σ modifier deltas)


class ReceiptPayment(BaseModel):
    method: PaymentMethod
    amount: Decimal
    reference: str | None
    tendered_amount: Decimal | None


class ReceiptRead(BaseModel):
    store: ReceiptStore
    order_number: str
    status: OrderStatus
    channel: OrderChannel
    table_number: str | None
    cashier_name: str | None
    customer_name: str | None
    created_at: datetime
    closed_at: datetime | None
    currency: str

    lines: list[ReceiptLine]

    # ── Totals (snapshot from the Order) ────────────────────────────
    subtotal: Decimal
    discount_total: Decimal
    service_charge: Decimal
    service_charge_rate: Decimal
    tax_total: Decimal
    tax_rate: Decimal
    tax_inclusive: bool
    tip_total: Decimal
    rounding_adjustment: Decimal
    total: Decimal
    paid_total: Decimal
    change_due: Decimal

    payments: list[ReceiptPayment]
    footer: str | None

File: frontend/src/types/receipt.ts
import type { OrderChannel, OrderStatus } from '@/types/order'
import type { PaymentMethod } from '@/types/payment'

/** Receipt DTO from GET /orders/{id}/receipt (M11). Decimals arrive as strings. */
export interface ReceiptStore {
  name: string
  address: string | null
  tax_id: string | null
}

export interface ReceiptModifier {
  name: string
  price_delta: string
}

export interface ReceiptLine {
  product_name: string
  qty: number
  unit_price: string
  modifiers: ReceiptModifier[]
  line_total: string
}

export interface ReceiptPayment {
  method: PaymentMethod
  amount: string
  reference: string | null
  tendered_amount: string | null
}

export interface Receipt {
  store: ReceiptStore
  order_number: string
  status: OrderStatus
  channel: OrderChannel
  table_number: string | null
  cashier_name: string | null
  customer_name: string | null
  created_at: string
  closed_at: string | null
  currency: string
  lines: ReceiptLine[]
  subtotal: string
  discount_total: string
  service_charge: string
  service_charge_rate: string
  tax_total: string
  tax_rate: string
  tax_inclusive: boolean
  tip_total: string
  rounding_adjustment: string
  total: string
  paid_total: string
  change_due: string
  payments: ReceiptPayment[]
  footer: string | null
}
